halo/evmstaking/evmstaking.go

```diff
@@ -0,0 +1,177 @@
+	address  common.Address
+	sKeeper  *skeeper.Keeper
+	bKeeper  bkeeper.Keeper
+	aKeeper  akeeper.AccountKeeper
+}
+
+// New returns a new LogProvider.
+func New(
+	ethCl ethclient.Client,
+	sKeeper *skeeper.Keeper,
+	bKeeper bkeeper.Keeper,
+	aKeeper akeeper.AccountKeeper,
+) (LogProvider, error) {
+	address := common.HexToAddress(predeploys.OmniStake)
+	contract, err := bindings.NewOmniStake(address, ethCl)
+	if err != nil {
+		return LogProvider{}, errors.Wrap(err, "new omni stake")
+	}
+
+	return LogProvider{
+		contract: contract,
+		ethCl:    ethCl,
+		address:  address,
+		sKeeper:  sKeeper,
+		bKeeper:  bKeeper,
+		aKeeper:  aKeeper,
+	}, nil
+}
+
+// Logs returns all omni stake contract EVM logs from the provided block hash.
+func (l LogProvider) Logs(ctx context.Context, blockHash common.Hash) ([]*evmenginetypes.EVMLog, error) {
+	// TODO(corver): Maybe we should filter by expected topic only.
+	logs, err := l.ethCl.FilterLogs(ctx, ethereum.FilterQuery{
+		BlockHash: &blockHash,
+		Addresses: l.Addresses(),
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "filter logs")
+	}
+
+	resp := make([]*evmenginetypes.EVMLog, 0, len(logs))
+	for _, l := range logs {
+		topics := make([][]byte, 0, len(l.Topics))
+		for _, t := range l.Topics {
+			topics = append(topics, t.Bytes())
+		}
+		resp = append(resp, &evmenginetypes.EVMLog{
+			Address: l.Address.Bytes(),
+			Topics:  topics,
+			Data:    l.Data,
+		})
+	}
+
+	return resp, nil
+}
+
+func (l LogProvider) Addresses() []common.Address {
+	return []common.Address{l.address}
+}
+
+// DeliverLog processes a omni deposit log event:
+// - Mint the corresponding amount of $STAKE coins.
+// - Send the minted coins to the depositor's account.
+// - Create a new validator with the depositor's account.
+func (l LogProvider) DeliverLog(ctx context.Context, _ common.Hash, elog *evmenginetypes.EVMLog) error {
+	deposit, err := l.contract.ParseDeposit(elog.ToEthLog())
+	if err != nil {
+		return errors.Wrap(err, "parse deposit")
+	}
+
+	log.Info(ctx, "EVM staking deposit detected, adding new validator",
+		"depositor", deposit.Depositor.Hex(),
+		"amount", deposit.Amount.String())
+
+	// TODO(corver): Replace this dummy pubkey with real pubkey from the deposit event once added.
+	dummyPubKey := k1.GenPrivKeySecp256k1(deposit.Depositor.Bytes()).PubKey()
+	pubkey, err := k1util.PubKeyToCosmos(dummyPubKey)
+	if err != nil {
+		return errors.Wrap(err, "pubkey to cosmos")
+	}
+
+	accAddr := sdk.AccAddress(deposit.Depositor.Bytes())
+	valAddr := sdk.ValAddress(deposit.Depositor.Bytes())
+	consAddr := sdk.GetConsAddress(pubkey)
+
+	if _, err := l.sKeeper.GetValidator(ctx, valAddr); err == nil {
+		return errors.New("validator already exists")
+	} else if _, err := l.sKeeper.GetValidatorByConsAddr(ctx, consAddr); err == nil {
+		return errors.New("validator pubkey already exists")
+	}
+
+	amountCoin, amountCoins := omniToBondCoin(deposit.Amount)
+
+	_ = l.getOrCreateAccount(ctx, accAddr)
```

Comment:
So if I understand correctly, if someone stakes with Omni we create an an account and a validator. What happens if they withdraw their stake? Do we have to call [RemoveAccount](https://pkg.go.dev/github.com/cosmos/cosmos-sdk/x/auth/keeper#AccountKeeper.RemoveAccount)?

Reply:
no, I think we can keep the account forever

Reply:
Do we only set the validator power to 0 or something like that?

Reply:
yup